spring-integration-core/src/main/java/org/springframework/integration/config/SmartLifecycleRoleController.java

```diff
@@ -0,0 +1,194 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.integration.config;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map.Entry;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.NoSuchBeanDefinitionException;
+import org.springframework.cloud.cluster.leader.event.AbstractLeaderEvent;
+import org.springframework.cloud.cluster.leader.event.OnGrantedEvent;
+import org.springframework.cloud.cluster.leader.event.OnRevokedEvent;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+import org.springframework.context.ApplicationListener;
+import org.springframework.context.SmartLifecycle;
+import org.springframework.util.Assert;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+
+/**
+ * Bulk start/stop {@link SmartLifecycle} in a particular role in phase order.
+ *
+ * @author Gary Russell
+ * @since 4.2
+ *
+ */
+public class SmartLifecycleRoleController implements ApplicationListener<AbstractLeaderEvent>,
+		ApplicationContextAware {
+
+	private static final Log logger = LogFactory.getLog(SmartLifecycleRoleController.class);
+
+	private final MultiValueMap<String, SmartLifecycle> lifecycles = new LinkedMultiValueMap<String, SmartLifecycle>();
+
+	private final MultiValueMap<String, String> lazyLifecycles = new LinkedMultiValueMap<String, String>();
+
+	private ApplicationContext applicationContext;
+
+	/**
+	 * Construct an instance with the provided lists of roles and lifecycles, which must be of equal length.
+	 * @param roles the roles.
+	 * @param lifecycles the lifecycles corresponding to the roles.
+	 */
+	public SmartLifecycleRoleController(List<String> roles, List<SmartLifecycle> lifecycles) {
+		Assert.notNull(roles, "'roles' cannot be null");
+		Assert.notNull(lifecycles, "'lifecycles' cannot be null");
+		Assert.isTrue(roles.size() == lifecycles.size(), "'roles' and 'lifecycles' must be the same lenght");
+		Iterator<SmartLifecycle> iterator = lifecycles.iterator();
+		for (String role : roles) {
+			SmartLifecycle lifecycle = iterator.next();
+			addLifecycleToRole(role, lifecycle);
+		}
+		for (List<SmartLifecycle> smarts : this.lifecycles.values()) {
+			Collections.sort(smarts, new Comparator<SmartLifecycle>() {
```

Comment:
This sort does not make sense here, because we can do `addLifecycleToRole` afterwards.
So, it has to be moved to the `startLifecyclesInRole`